[tf-psa-crypto] Align header include guards with their new locations in TF PSA Crypto

valeriosetti opened the pull request:
## Description

Resolves #399

Due to the large amount of files affected and the repetitive type of work, all the changes in this PR were generated by a Python script. This means that in case of any conflict with ongoing works, this PR can easily be updated.

The script is currently included in the 1st commit of this PR for reviewers' reference. I put it in `./scripts/` folder and I designed it in a way that can be used both for updating guards and also for checking them. Currently this check feature is not included in the CI, but I added that because I thought it could be a nice feature to have in the future.

## PR checklist

- [x] **changelog** not required because: this change should not affect end user 
- [x] **framework PR** ~provided https://github.com/Mbed-TLS/mbedtls-framework/pull/216~ not required
- [x] **mbedtls development PR** not required because: no change there
- [x] **mbedtls 3.6 PR** not required because: no backport
- **tests** not required because: no code change

mpg commented:
> Because they are very indented in the tree and I feared that this could cause some code style complain for going over the 100 columns.

Actually the C standard only requires the 63 first characters to be significant in a macro name, so I guess we should avoid macro names longer that 63 chars? Or are we confident that all implementations used in practice will not have this limit? (Unsurprisingly, [GCC has no limit](https://gcc.gnu.org/onlinedocs/cpp/Implementation-limits.html) and I'd expect the same from Clang.)

gilles-peskine-arm commented:
> For the time being build_psa_alt_headers is the only failing test in the CI, but this requires a fix in the mbedtls branch.

I haven't looked closely, but at a guess, the problem is that both projects have headers with the name name? That's a move in progress, the crypto files have already been created but [the mbedtls files are still there](https://github.com/Mbed-TLS/mbedtls/pull/10394). I wasn't prioritizing the removal because it seemed like a minor cleanup, but I'll raise the priority it since it's blocking something else. (Still below 1.0 work though.)

valeriosetti commented:
> > For the time being build_psa_alt_headers is the only failing test in the CI, but this requires a fix in the mbedtls branch.
> 
> I haven't looked closely, but at a guess, the problem is that both projects have headers with the name name? That's a move in progress, the crypto files have already been created but [the mbedtls files are still there](https://github.com/Mbed-TLS/mbedtls/pull/10394). I wasn't prioritizing the removal because it seemed like a minor cleanup, but I'll raise the priority it since it's blocking something else. (Still below 1.0 work though.)

Actually I think the problem is related to the fact `_alt.h` header files are generated form the original ones with:
`make -C tests ../framework/tests/include/alt-extra/psa/crypto_platform_alt.h ../framework/tests/include/alt-extra/psa/crypto_struct_alt.h`
and in the original ones I renamed `PSA_CRYPTO_PLATFORM_H` -> `TF_PSA_CRYPTO_INCLUDE_PSA_CRYPTO_PLATFORM_H`, but at the same time checks in `framework/tests/programs/query_included_headers.c` are still in the "old" format.

gilles-peskine-arm commented:
> If so, what about `TF_PSA_CRYPTO_<rel_path>_<filename>_H` where `rel_path` is the path through which the header file is usually included in the library (ex: `mbedtls/xxx.h`)?

That is what I would have expected a priori.

valeriosetti committed:
scripts: add a script to automatically check/fix guards in C header files

This will be used in the following commit to update private header files in the
tf-psa-crypto library.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on drivers/builtin/include/mbedtls/config_adjust_legacy_crypto.h:
```diff
@@ -21,8 +21,8 @@
-#ifndef MBEDTLS_CONFIG_ADJUST_LEGACY_CRYPTO_H
-#define MBEDTLS_CONFIG_ADJUST_LEGACY_CRYPTO_H
+#ifndef TF_PSA_CRYPTO_MBEDTLS_CONFIG_ADJUST_LEGACY_CRYPTO_H
+#define TF_PSA_CRYPTO_MBEDTLS_CONFIG_ADJUST_LEGACY_CRYPTO_H
```
I thought this was the [pattern we agreed on](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/460#issuecomment-3302103636) for these guards, but apparently I missed something. Didn't we want a `TF_PSA_CRYPTO_` prefix for all the guards?

valeriosetti committed:
drivers|include: fix guards of private include files

Pattern is as follows: TF_PSA_CRYPTO_<path>_<filename>_H
This was automatically generated by python script 'script/c_header_guards.py'.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg commented:
Unfortunately this now has conflicts. Can you rebase? Thanks!

valeriosetti commented:
> Unfortunately this now has conflicts. Can you rebase? Thanks!

Yeah, I knew this, but as I mentioned to Ben yesterday on Slack since all the changes in this PR are automated through a Python script (so it's easy to re-run) and at the same it's easy to get conflicts, my preferred workflow was:
1. find an overall agreement on the proposed implementation;
2. rebase and fix conflicts;
3. re-approve shortly after the rebase in order to avoid conflicts.

@mpg if you agree with current design I will proceed with step (2)

mpg commented:
Ah yes, makes sense! I should be able to review later this morning.

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+    TF_PSA_CRYPTO_<rel_path>_<filename>_H
+where:
+- 'TF_PSA_CRYPTO_' is the common hardcoded prefix;
+- 'rel_path' is the path used in C files to include such header file.
```
I don't think that's what the script implements. I think for public headers it's just `<rel_path>_<filename>_H` and the prefix is only added for private headers.

Note: I think that's also not exactly the rule [Gilles was suggesting](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/460#issuecomment-3311375010): my understanding was that (even for private headers) we would only add a `TF_PSA_CRYPTO_` prefix if `<rel_path>` doesn't already start with one of `MBEDTLS_`, `PSA_` or `TF_PSA_CRYPTO`. Note that for all public headers, `rel_path` always starts with one of these, and for private headers sometimes it does (`drivers/builtin/include`) and sometimes not (`drivers/builtin/src`).

I'm not sure if the deviation from Gilles's suggestion was intentional?

Regardless we want the documentation to be aligned with the code :)

valeriosetti replied:
> I'm not sure if the deviation from Gilles's suggestion was intentional?

No, I missed that part. OTOH since all we're changing here is guards for the header files, wouldn't it be better to use a fixed and simple approach (like the one that you just described in the 1st paragraph of your comment) instead of adding exceptions that make code harder to read?

> Regardless we want the documentation to be aligned with the code :)

Yes, I totally forgot to update it and I will do it now.

valeriosetti replied:
I included the "core/" folder in the search and check path of the script. The final result is at [this commit](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/460/commits/940d834416388ce492806ac11de3ddf7ce9c472a).
Is it a reasonable result in your opinion?

I know that for some header files which already have `psa_crypto_` in the file name (ex: `core/psa_crypto_core.h`) the final result is a bit too long (for the file that I just mentioned it is `TF_PSA_CRYPTO_PSA_CRYPTO_CORE_H`), but I prefer to have a common pattern, rather than making exceptions. However I know that this is a matter of preference :)

mpg commented on drivers/builtin/include/mbedtls/private/aes.h:
```diff
@@ -25,8 +25,8 @@
 
-#ifndef MBEDTLS_AES_H
-#define MBEDTLS_AES_H
+#ifndef TF_PSA_CRYPTO_MBEDTLS_PRIVATE_AES_H
```
For example I think `MBEDTLS_PRIVATE_AES_H` would be good enough here.

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+# - key: path of header files to check;
+# - value: path that must be trimmed from the key value in order to get the
+#          final guard.
+FOLDERS_TO_CHECK = {
```
I'm also seeing headers in `core`.

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+    adjusted_name = re.sub(r'[^a-zA-Z0-9]', '_', adjusted_name).upper()
+    guard_elements = []
+    # Add prefix only to non public header files
+    if file_path.startswith('drivers'):
```
Missing private headers in `core`.

Also, if we switch to the rule suggested by Gilles, the condition should be
```py
if not (adjusted_name.startswith('MBEDTLS_') or adjusted_name.startswith('PSA_') or adjusted_name.startswith('TF_PSA_CRYPTO_')):
```
That is, we no longer need to distinguish between private and public headers, we just check if the current name already has one of our 3 prefixes or not.

valeriosetti replied:
Let's find an agreement on the [pattern](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/460#discussion_r2454888334) first and then if we agree to go this way I will update the script ;)

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+- 'TF_PSA_CRYPTO_' is the common hardcoded prefix;
+- 'rel_path' is the path used in C files to include such header file.
+
+Usage: python3 fix_header_guards.py
```
The file is called `c_header_guards.py` now :)

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+
+def find_line(lines: List[str], search_range, regex: str) -> Tuple[int, Match]:
+    """Try to match the regex expression on the range of lines given in input.
+    The first successful match is returned; an execption is returned if no
```
Typo: exception

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+    On failure an exception is thrown.
+    """
+    with open(file_path, 'r', encoding='utf-8') as f:
+        content = f.read()
```
Minor: a bit more compact than this plus `lines = content.split('\n')` below would be `lines = f.readlines()` here.

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+
+    Input params:
+        lines: content of the file given as list of strings;
+        search_range: range() to use when going through the list of lines;
```
Note: I like how you're using a range object as a parameter here, that's not something I would have considered, I've learned a new useful Python idiom today, thanks!

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+        if fix:
+            lines[ifndef_line] = expected_line
+        else:
+            raise ValueError(f'Invalid guard format:\n'
```
Note: I'm not a big fan of using exceptions for normal errors, but I guess it's a matter of taste/style, and considering this script is not central, I'm not going to argue about this.

mpg commented on scripts/c_header_guards.py:
```diff
@@ -0,0 +1,157 @@
+    current_path = os.getcwd()
+
+    arg_parser = argparse.ArgumentParser()
+    arg_parser.add_argument('--fix', action='store_true',
```
Ah, so there is a command-line option. That should probably be reflected in the "Usage" line in the script's docstring.

mpg reviewed (changes requested):
Looking pretty good, most of the few issues I found are mostly in docstring, I think the only one that matters in the code is that `core` is missing.

Also, since the script can check guards, not just fix them, I think the check should be added to `all.sh`, to catch any mistakes in future header files we create. Probably a new component in `tests/scripts/components-basic-checks.sh`.

valeriosetti committed:
scripts: c_header_guards.py: check also "core/" folder

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
core: fix guards for header files in "core/" folder

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
scripts: c_header_guards.py: fix script's docstring and another typo

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
scripts: c_header_guards.py: optimize lines reading from header file

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
scripts: c_header_guards.py: fix header files match regex

Match only files which are ending with ".h" not those which have ".h" in
the middle.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: components-basic-checks: add new component for header files guards

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (approved):
LGTM, thanks!
